src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java

```diff
@@ -0,0 +1,110 @@
+
+    private final SubscriptionRepository subscriptionRepository;
+
+    public SubscriptionSyncController(SubscriptionRepository subscriptionRepository) {
+        this.subscriptionRepository = subscriptionRepository;
+    }
+
+    public void syncSubscription(Subscription subscription) {
+        final Optional<org.candlepin.subscriptions.db.model.Subscription> entity =
+            subscriptionRepository.findActiveSubscription(String.valueOf(subscription.getId()));
+        if (entity.isPresent()) {
+            final org.candlepin.subscriptions.db.model.Subscription dbSub = entity.get();
+            if (updateSubscription(subscription, dbSub)) {
+                // end the current record
+                dbSub.setEndDate(OffsetDateTime.now());
+                subscriptionRepository.save(dbSub);
+                // create a new record
+                final org.candlepin.subscriptions.db.model.Subscription newSub =
+                    new org.candlepin.subscriptions.db.model.Subscription();
+                newSub.setSubscriptionId(dbSub.getSubscriptionId());
+                newSub.setSku(dbSub.getSku());
+                newSub.setOwnerId(dbSub.getOwnerId());
+                newSub.setQuantity(subscription.getQuantity());
+                newSub.setStartDate(OffsetDateTime.now());
+                if (subscription.getEffectiveEndDate() != null) {
+                    newSub.setEndDate(convertLongTime(subscription.getEffectiveEndDate()));
+                }
+                subscriptionRepository.save(newSub);
+            }
+            else {
+                subscriptionRepository.save(dbSub);
+            }
+        }
+        else {
+            // create a new record
+            final org.candlepin.subscriptions.db.model.Subscription newSub =
+                new org.candlepin.subscriptions.db.model.Subscription();
+            newSub.setSubscriptionId(String.valueOf(subscription.getId()));
+            newSub.setQuantity(subscription.getQuantity());
+            if (subscription.getEffectiveStartDate() != null) {
+                newSub.setStartDate(convertLongTime(subscription.getEffectiveStartDate()));
+            }
+            if (subscription.getEffectiveEndDate() != null) {
+                newSub.setEndDate(convertLongTime(subscription.getEffectiveEndDate()));
+            }
+            subscriptionRepository.save(newSub);
+        }
+    }
+
+    private static boolean updateSubscription(Subscription dto,
+        org.candlepin.subscriptions.db.model.Subscription entity) {
+        if (dto.getQuantity() != entity.getQuantity()) {
+            return true;
+        }
+        if (dto.getEffectiveStartDate() != null) {
+            entity.setStartDate(
+                convertLongTime(dto.getEffectiveStartDate())
+            );
+        }
+        if (dto.getEffectiveEndDate() != null) {
+            entity.setEndDate(
+                convertLongTime(dto.getEffectiveEndDate())
+            );
+        }
```

Comment:
Was thinking about this, and I think we need to keep start date out of this, because imagine the following scenario:

1. Subscription state is synced:
```yaml
start: 2020-01-01
end: 2020-12-31
quantity: 10
```
-> creates a record that looks like that.
2. Subscription state changes to
```yaml

start: 2020-01-01
end: 2020-12-31
quantity: 5
```
-> saves as two subscription records, one with endDate > now()
3. sync is called again with the same data, active subscription has start != 2020-01-01... -> a third subscription record created...

It'd be ideal to unit test this scenario.

Reply:
So would it instead be reasonable to set the end date to now and create a new record with the updated info if quantity or start date changes? That way our "active data" is only the most current information we have.

Reply:
Actually, I guess similar issue happens with start date, so we might have to drop both from the check.